File: lib/features/auth/data/models/user_model.dart
class UserModel {
  const UserModel({
    required this.id,
    required this.fullName,
    required this.email,
    this.publicUserId,
    this.localUserId,
    this.displayName,
    this.emergencyNote,
    this.phoneNumber,
    this.avatarUrl,
    this.role = 'user',
  });

  final String id;
  final String fullName;
  final String email;
  final String? publicUserId;
  final String? localUserId;
  final String? displayName;
  final String? emergencyNote;
  final String? phoneNumber;
  final String? avatarUrl;
  final String role;

  factory UserModel.fromJson(Map<String, dynamic> json) {
    final displayName = json['displayName']?.toString();
    final fullName = (json['fullName'] ?? displayName ?? '').toString();
    return UserModel(
      id: (json['id'] ?? json['_id']).toString(),
      fullName: fullName,
      email: json['email']?.toString() ?? '',
      publicUserId: json['publicUserId']?.toString(),
      localUserId: json['localUserId']?.toString(),
      displayName: displayName,
      emergencyNote: json['emergencyNote']?.toString(),
      phoneNumber: json['phoneNumber']?.toString(),
      avatarUrl: json['avatarUrl']?.toString(),
      role: json['role']?.toString() ?? 'user',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'fullName': fullName,
      'email': email,
      'publicUserId': publicUserId,
      'localUserId': localUserId,
      'displayName': displayName,
      'emergencyNote': emergencyNote,
      'phoneNumber': phoneNumber,
      'avatarUrl': avatarUrl,
      'role': role,
    };
  }
}

File: backend/src/models/user.model.js
const mongoose = require('mongoose');

const userSchema = new mongoose.Schema(
  {
    publicUserId: {
      type: String,
      trim: true,
    },
    localUserId: {
      type: String,
      trim: true,
    },
    fullName: {
      type: String,
      required: true,
      trim: true,
    },
    displayName: {
      type: String,
      trim: true,
      default: null,
    },
    email: {
      type: String,
      lowercase: true,
      trim: true,
    },
    passwordHash: {
      type: String,
      required: true,
    },
    authProvider: {
      type: String,
      enum: ['password', 'bootstrap'],
      default: 'password',
    },
    isBootstrapProfile: {
      type: Boolean,
      default: false,
    },
    bootstrapLocalUserId: {
      type: String,
      trim: true,
      default: null,
      index: true,
    },
    emergencyNote: {
      type: String,
      trim: true,
      default: null,
    },
    identitySource: {
      type: String,
      enum: ['local_bootstrap', 'direct_cloud'],
      default: 'direct_cloud',
    },
    createdFromDeviceId: {
      type: String,
      trim: true,
      default: null,
    },
    phoneNumber: {
      type: String,
      trim: true,
      default: null,
    },
    avatarUrl: {
      type: String,
      trim: true,
      default: null,
    },
    role: {
      type: String,
      enum: ['user', 'admin'],
      default: 'user',
    },
    lastLoginAt: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true,
  },
);

userSchema.index(
  { publicUserId: 1 },
  {
    unique: true,
    partialFilterExpression: { publicUserId: { $type: 'string' } },
  },
);
userSchema.index(
  { localUserId: 1 },
  {
    unique: true,
    partialFilterExpression: { localUserId: { $type: 'string' } },
  },
);
userSchema.index(
  { email: 1 },
  {
    unique: true,
    partialFilterExpression: { email: { $type: 'string' } },
  },
);

module.exports = mongoose.model('User', userSchema);
